source/h1_connection.c

```diff
@@ -1294,33 +1461,58 @@ static int s_handler_process_read_message(
             goto shutdown;
         }
 
-        if (!connection->thread_data.incoming_stream) {
-            /* TODO: Server connection would create new request-handler stream at this point. */
+        /* When connection has switched protocols, messages are processed very differently.
+         * They're queued and sent along whenever the downstream read-window can accommodate them.
+         *
+         * We need to do this check in the middle of the normal processing loop,
+         * in case the switch happens in the middle of processing a message. */
+        if (connection->thread_data.has_switched_protocols) {
+            /* Don't auto-increment read window for parts of message using the new protocol. */
+            if (message_cursor.len < connection->thread_data.incoming_message_window_update) {
+                connection->thread_data.incoming_message_window_update -= message_cursor.len;
+            } else {
+                connection->thread_data.incoming_message_window_update = 0;
+            }
 
```

Comment:
It took me longer than I want to document publicly to figure this out. This is setting copy_mark to just after the current message_cursor's worth of data in the message, is that correct? Could you comment that for future us?

Reply:
yes

Reply:
I changed the auto-window-update stuff in an attempt to make it less confusing